### PR TITLE
Update world.md

### PR DESCRIPTION
**Pull request by michaelm-rsi**
First and very early draft.

# Description

Expand the documentation on the World class to explain how to use it to more fully leverage code reuse.

# Motivation & context

Documentation improvement, and adding a bit of a tutorial.

## Type of change

- Documentation

# Checklist:

- [x ] I have read the [**CONTRIBUTING**](../CONTRIBUTING.md) document.
- [ ] My change needed additional tests
  - [ ] I have added tests to cover my changes.
- [ x] My change requires a change to the documentation.
  - [x ] I have updated the documentation accordingly.
- [ ] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../CHANGELOG.md), linking to this pull request.

### CODE REVIEW

File: docs/support_files/world.md [hunk 1]
```diff
@@ -1,32 +1,47 @@
...
-* `this.attach`: function used for adding [attachments](./attachments.md) to hooks/steps
-* `this.log`: function used for [logging](./attachments.md#logging) information from hooks/steps
-* `this.parameters`: object of parameters passed in via the [CLI](../cli.md#world-parameters)
+## Cucumber World
 
-Some notes on the scope of World:
+The default world provided by Cucumber provides a number of formatting helpers that are passed into the constructor:
 
-- It's scoped to a single scenario only - not shared globally between scenarios. This reinforces a Cucumber principle: that scenarios should work entirely independently of one another.
-- If you're using [Retry](../retry.md), you'll get a fresh World for every attempt at your scenario, so state isn't retained between attempts.
```
Comment (davidjgoss) [inline docs/support_files/world.md:27]: We've lost any reference to Retry now but this is an important detail.

File: docs/support_files/world.md [hunk 2]
```diff
@@ -36,32 +51,85 @@ $ cucumber-js --world-parameters '{"appUrl":"http://localhost:3000/"}'
...
-        return await this.driver.wait(condition)
-    }
+import { Given } from '@cucumber/cucumber';
+import CustomWorld from "../classes/CustomWorld.js"
+
+Given("I am an {string}", function(userType) {
+  this.login(userType)
+})
+
+CustomWorld.prototype.login(userType) {
```
Comment (davidjgoss) [inline docs/support_files/world.md:118]: This is one possible pattern of several, and it's beyond the scope of this document to be recommending any particular one - would be an interesting blog post though!

Here, I think it's enough to demonstrate methods being called from steps/hooks on a custom world, and noting that as your world does more you'll need to think about how to manage that.

File: docs/support_files/world.md [hunk 3]
```diff
@@ -1,32 +1,47 @@
...
+  Then my color should not be red  
+```
 
-* `this.attach`: function used for adding [attachments](./attachments.md) to hooks/steps
-* `this.log`: function used for [logging](./attachments.md#logging) information from hooks/steps
-* `this.parameters`: object of parameters passed in via the [CLI](../cli.md#world-parameters)
+## Cucumber World
 
-Some notes on the scope of World:
+The default world provided by Cucumber provides a number of formatting helpers that are passed into the constructor:
```
Comment (davidjgoss) [inline docs/support_files/world.md:37]: These things are supplied to all World constructors including custom ones - the custom world example hints at this but I think it's worth being explicit.

File: docs/support_files/world.md [hunk 4]
```diff
@@ -36,32 +51,85 @@ $ cucumber-js --world-parameters '{"appUrl":"http://localhost:3000/"}'
...
+  constructor(options) {
+    super(options);
+    // Custom actions go here.
+  }
+
+  /*
+   * Constructors cannot be asynchronous! To work around this we'll
+   * use an init method with the Before hook
+   */
+  async init(scenario) {
```
Comment (davidjgoss) [inline docs/support_files/world.md:85]: Good improvement 👍

**Review by davidjgoss — CHANGES_REQUESTED** [review]

File: docs/support_files/world.md (hunk 1 above)
Comment (michaelm-rsi) [reply docs/support_files/world.md:27]: It's now back in in two locations - first paragraph and in the summary.

File: docs/support_files/world.md (hunk 4 above)
Comment (michaelm-rsi) [reply docs/support_files/world.md:114]: Thanks. Added a bit more, is it too much or good. I think the above at this point is all I can think of.

File: docs/support_files/world.md (hunk 1 above)
Comment (michaelm-rsi) [reply docs/support_files/world.md:27]: Added two references in, one in the first paragraph and another in the summary.

File: docs/support_files/world.md (hunk 2 above)
Comment (michaelm-rsi) [reply docs/support_files/world.md]: Agreed and removed.

File: docs/support_files/world.md (hunk 3 above)
Comment (michaelm-rsi) [reply docs/support_files/world.md]: Added. Also added the structure of the options object in the comments of the custom world code example.

File: docs/support_files/world.md (hunk 4 above)
Comment (michaelm-rsi) [reply docs/support_files/world.md:114]: Thanks.

**Review by aurelien-reeves — APPROVED** [review]
I definitely love your PR!
Just waiting for @davidjgoss (definitive?) review 👌

**Review by davidjgoss — APPROVED** [review]

**Pull request merged**